File: swara-learn/src/main/java/com/swara/ml/neural/layers/ForwardLayer.java
package com.swara.ml.neural.layers;

import java.util.Stack;

import com.swara.ml.neural.NeuralLayer;

import org.apache.commons.math3.analysis.differentiation.DerivativeStructure;
import org.apache.commons.math3.analysis.differentiation.UnivariateDifferentiableFunction;
import org.apache.commons.math3.analysis.function.Logistic;
import org.apache.commons.math3.linear.Array2DRowRealMatrix;
import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.RealVector;

/**
 * A feed forward neural layer with a parameterizable activation function. Feed forward layers are
 * fully-connected; every neuron in the layer is connected to the output of every layer in the
 * previous layer. The weights of forward layers are updated via simple backpropagation.
 */
public class ForwardLayer extends NeuralLayer {

    private final UnivariateDifferentiableFunction activation;
    private final Stack<RealVector> history;
    private RealMatrix weights;

    public ForwardLayer(int inputs, int outputs) {
        // By default, construct a sigmoid function with a 0.5 learn rate.
        this(inputs, outputs, new Logistic(1, 1, 0.5, 1, 0, 1));
    }

    public ForwardLayer(int inputs, int outputs, UnivariateDifferentiableFunction activation) {
        super(inputs, outputs);
        this.activation = activation;
        this.history = new Stack<>();
        this.weights = new Array2DRowRealMatrix(outputs, inputs + 1);
    }

    @Override
    public RealVector forward(RealVector input) {
        // Append a bias weight to the input, and calculate the linear combination.
        final RealVector netj = this.history.push(this.weights.operate(input.append(1.0)));
        return this.history.push(netj.map(this.activation));
    }

    @Override
    public RealVector backward(RealVector error) {
        final RealVector output = this.history.pop();
        final RealVector input  = this.history.pop();

        // Calculate the gradient of the activation function with respect to the last input to the
        // activation (netj in wikipedia description).
        final RealVector gradient = error.ebeMultiply(input.map(i -> this.activation
            .value(new DerivativeStructure(1, 1, 0, i))
            .getPartialDerivative(1)
        ));

        // Calculate the error that is propagated to the previous layer.
        final RealVector next = this.weights.preMultiply(gradient);
        this.weights = this.weights.add(gradient.outerProduct(output));
        return next;
    }

}

File: swara-learn/src/main/scala/com/swara/learn/neural/layers/ForwardLayer.scala
package com.swara.learn.neural.layers

import breeze.numerics
import breeze.optimize.DiffFunction
import breeze.stats.distributions.Rand
import com.swara.learn.neural.Layer

/**
 * A feed-forward, fully-connected layer. A forward layer consists of a weight matrix, W, in which
 * W(i, j) indicates the weight of the ith neuron for the jth input, a bias vector, B, in which B(k)
 * indicates the bias of the kth neuron for all inputs, and a monotonically increasing, bounded,
 * differentiable activation function, A. The forward layer is essentially a linear combination
 * machine; for some input vector x, the output of the forward layer is A(Wx + B).
 *
 * @param activation Monotonically increasing, bounded, differentiable function.
 * @param weights Weight matrix.
 * @param biases Bias vector.
 */
class ForwardLayer(
  activation: DiffFunction[Double],
  weights: Matrix,
  biases: Vector
) extends Layer[Vector, Vector] {

  require(this.weights.rows == this.biases.length)

  override def apply(inputs: Seq[Vector]): Result = {
    // Calculate the weighted output of each neuron. (Wx + b)
    val weighted = inputs.map(this.weights * _ + this.biases)
    
    Result(weighted.map(_.map(this.activation)), { errors =>
      (inputs, errors, weighted).zipped.map { case (in, err, out) =>
        // Calculate the gradient for each neuron as the derivative of the activation function at
        // its weighted output, multipled by the error (Chain Rule).
        val gradients = out.map(this.activation.gradientAt) :* err

        // Update the weights/biases of the layer using gradient.
        this.biases -= gradients
        this.weights -= gradients.asDenseMatrix.t * in.asDenseMatrix

        // Backpropagate error to the previous layer.
        this.weights.t * gradients
      }
    })
  }


//  def forward(x: Vector): Vector =
//    (this.weights * x + this.biases).map(this.activation)
//
//  def backward(examples: Seq[(Vector, Vector, Vector)]): Seq[Vector] = {
//    // Calculate the gradients for each neuron, weight updates, and propagated error.
//    val (gradient, updates, propagated) = examples.map { case (input, _, error) =>
//      val gradient = (this.weights * input + this.biases).map(activation.gradientAt) :* error
//      val updates = gradient.asDenseMatrix.t * input.toDenseMatrix
//      val propagated = this.weights.t * gradient
//      (gradient, updates, propagated)
//    }.unzip3
//
//    // Update biases and weights using the calculated neuron gradients and weight updates.
//    this.biases -= gradient
//    this.weights -= updates
//
//    // Propagate the calculated error to the previous layer.
//    propagated
//  }

}

object ForwardLayer {

  /**
   * Constructs a forward layer that accepts the specified number of inputs and produces the
   * specified number of outputs. The initial weights and biases are uniformly random numbers
   * normalized using the equation described in http://stats.stackexchange.com/a/186351.
   *
   * @param activation Monotically increasing, bounded, differentiable function.
   * @param in Number of inputs.
   * @param out Number of outputs.
   */
  def apply(activation: DiffFunction[Double])(in: Int, out: Int): ForwardLayer = {
    val dist = Rand.uniform.map(x => (x * 2 - 1) / numerics.sqrt(in))
    new ForwardLayer(
      activation,
      Matrix.rand(out, in, dist),
      Vector.rand(out, dist)
    )
  }

  def identity: (Int, Int) => ForwardLayer =
    apply(new DiffFunction[Double] {
      override def calculate(x: Double): (Double, Double) = (x, 1)
    })

  def rectifier: (Int, Int) => ForwardLayer =
    apply(new DiffFunction[Double] {
      override def calculate(x: Double): (Double, Double) = if (x < 0) (0, 0) else (x, 1)
    })

  def sigmoid: (Int, Int) => ForwardLayer =
    apply(new DiffFunction[Double] {
      override def calculate(x: Double): (Double, Double) = {
        val fx = numerics.sigmoid(x)
        (fx, fx * (1 - fx))
      }
    })

  def tanh: (Int, Int) => ForwardLayer =
    apply(new DiffFunction[Double] {
      override def calculate(x: Double): (Double, Double) = {
        val fx = numerics.tanh(x)
        (fx, 1 - fx * fx)
      }
    })

}
